docker/DOCKER.md

```diff
@@ -0,0 +1,400 @@
+- `-e SSH_AUTH_SOCK`: share your host SSH agent socket with preloaded SSH keys (requires `-v "$(dirname "${SSH_AUTH_SOCK}")"`)
+- `-e DOCKERD=1`: enable the included Docker-in-Docker daemon (requires `--cap-add SYS_ADMIN`)
+
+## Usage
+
+We've provided some examples of common CLI commands and how they are best used
+with this image, since some special considerations must be made.
+
+- [login](#login) - login to balena
+- [push](#push) - start a build on the remote balenaCloud build servers, or a local mode device
+- [logs](#logs) - show device logs
+- [ssh](#ssh) - SSH into the host or application container of a device
+- [apps](#app--apps) - list all applications
+- [app](#app--apps) - display information about a single application
+- [devices](#device--devices) - list all devices
+- [device](#device--devices) - show info about a single device
+- [tunnel](#tunnel) - tunnel local ports to your balenaOS device
+- [preload](#preload) - preload an app on a disk image (or Edison zip archive)
+- [build](#build--deploy) - build a project locally
+- [deploy](#build--deploy) - deploy a single image or a multicontainer project to a balena application
+- [join](#join--leave) - move a local device to an application on another balena server
+- [leave](#join--leave) - remove a local device from its balena application
+- [scan](#scan) - scan for balenaOS devices on your local network
+
+For each example we have also linked to the corresponding sections of the
+balena CLI Documentation here: https://www.balena.io/docs/reference/balena-cli
+
+### login
+
+- <https://www.balena.io/docs/reference/balena-cli/#login>
+
+The `balena login` command can't be used with web authorization and a browser
+when running in a container. Instead it must be used with `--token` or `--credentials`.
+
+Notice that here we've used a named volume `balena_data` to store credentials
+for future runs of the CLI image. This is optional but avoids having to run the login
+command again every time you run the image.
+
+```bash
+$ docker volume create balena_data
+
+$ docker run --pull always --rm -it -v "balena_data:/root/.balena" \
+    balenalib/amd64-debian-balenacli login --credentials --email "johndoe@gmail.com" --password "secret"
+
+$ docker run --pull always --rm -it -v "balena_data:/root/.balena" \
+    balenalib/amd64-debian-balenacli login --token "..."
+```
+
+### push
+
+- <https://www.balena.io/docs/reference/balena-cli/#push-applicationordevice>
+
+In this example we are mounting your current working directory into the container with `-v "$PWD:$PWD" -w "$PWD"`.
+This will bind mount your current working directory into the container at the same absolute path.
+
+This bind mount is required so the CLI has access to your app sources.
+
+```bash
+$ docker run --pull always --rm -it -v "balena_data:/root/.balena" \
+    -v "$PWD:$PWD" -w "$PWD" \
+    balenalib/amd64-debian-balenacli bash
+
+> balena login --credentials --email "johndoe@gmail.com" --password "secret"
```

Comment:
I find the login command in every example to be really excessive. Do you think we can drop it and when users realize they aren't logged in they will scroll up to the corresponding login section?

Reply:
Yeah, I agree, no need to repeat the `balena login` line in several places. 👍

Reply:
The catch so far has been some balena commands are also valid executables in the system path, so just saying "is this cmd an executable file in PATH" will fall over with things like `ssh` and some others.

I'm trying to be too clever by supporting both args and commands in the CMD field, but I'm determined to make it work in a predictable way!

I was pretty close when I didn't check `[ -x "$(which "${1}") ]` and only checked `[ -x "${1} ]` but then it fails when providing `bash` instead of `/bin/bash`.

Is there a file in the source code where we could quickly grep whether something is a real balena command?

Reply:
Related, I'm going to limit the cases where the ssh-agent starts to only when the SSH_PRIVATE_KEY is provided and the agent is not already mapped from the host.